### PR TITLE
Adding package BDSim, default version v1.6.0

### PR DESCRIPTION
**Pull request by gganis**
Beam Delivery Simulation (BDSIM) is a C++ program that utilises the Geant4 toolkit
to simulate both the transport of particles in an accelerator and their interaction
with the accelerator material. BDSIM is capable of simulating a wide variety of
accelerator components, magnets and processes, e.g. synchrotron radiation, with Geant4
geometry dynamically built based on a text input file. Thick lens accelerator tracking
routines are provided for fast accurate tracking in a vacuum.

BEGINRELEASENOTES
- Beam Delivery Simulation (BDSIM) is a C++ program that utilises the Geant4 toolkit
to simulate both the transport of particles in an accelerator and their interaction
with the accelerator material. BDSIM is capable of simulating a wide variety of
accelerator components, magnets and processes, e.g. synchrotron radiation, with Geant4
geometry dynamically built based on a text input file. Thick lens accelerator tracking
routines are provided for fast accurate tracking in a vacuum.
- Current production version (v1.6.0) has been added. Note that i does not build (yet)
   against Geant4 11.2.0, so Geant4 10.7.3 has been enforced
ENDRELEASENOTES

### CODE REVIEW

**Comment by wdconinc** [comment]
If I may, I'd suggest this package is general enough for inclusion upstream.

File: packages/bdsim/package.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+# them, you can save this file and test your package like this:
+#
+#     spack install bdsim
+#
+# You can edit this file again by typing:
+#
+#     spack edit bdsim
+#
+# See the Spack documentation for more information on packaging.
+# ----------------------------------------------------------------------------
```
Comment (vvolkl) [inline packages/bdsim/package.py:21]:
```suggestion
```

File: packages/bdsim/package.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+from spack import *
+
+
+class Bdsim(CMakePackage):
+    """Beam Delivery Simulation (BDSIM) is a C++ program that utilises the Geant4 toolkit
+       to simulate both the transport of particles in an accelerator and their interaction
+       with the accelerator material"""
```
Comment (vvolkl) [inline packages/bdsim/package.py:13]:
```suggestion
       to simulate both the transport of particles in an accelerator and their 
       interaction with the accelerator material"""
```
In order to satisfy `spack style`

File: packages/bdsim/package.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    version('1.6.0', sha256='e3241d2d097cb4e22249e315c1474da9b3657b9c6893232d9f9e543a5323f717')
+
+    depends_on('cmake')
+    depends_on('geant4')
+    depends_on('geant4@:10.7.3', when='@:1.6.0')
+    depends_on('root')
+    depends_on('clhep')
+    depends_on('flex')
+    depends_on('bison')
+
```
Comment (vvolkl) [inline packages/bdsim/package.py:33]:
```suggestion
```
In order to satisfy `spack style`

**Comment by vvolkl** [comment]
When opening pull requests on upstream spack, flake8 is run, I've added two more suggestions to make this file pass the spack CI. Like this it can be added to spack/spack. I've already added it to the key4hep fork in order to include it in the next release

**Comment by gganis** [comment]
Many thanks @vvolkl, I have validated/commit all your suggestions.

**Comment by gganis** [comment]
Let me know when you believe we can merge

**Comment by vvolkl** [comment]
The bdsim recipe is now in our fork and in https://github.com/spack/spack/pull/31384, so here we just need to add it to the key4hep-stack bundle.

**Pull request merged**